<a href="https://colab.research.google.com/github/IGiovanniRamirez/Giovanni-Ramirez-Data-Portfolio/blob/main/RappiPlus%20de%20datos%20a%20decisiones%20de%20negocio/RappiPlus.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Proyecto RappiPlus: de datos a decisiones de negocio

**Introducción**


El objetivo de este proyecto es evaluar el desempeño del servicio **RappiPlus** para apoyar **decisiones de negocio basadas en datos**.

Se trabajan con múltiples datasets del negocio:

- **rappiplus_orders_raw.csv** → información de pedidos, precios, descuentos y revenue  
- **rappiplus_catalog.csv** → costos de productos, categorías y proveedores  
- **rappiplus_marketing_spend.csv** → inversión en marketing por canal y país  
- **events / users / user_activity (SQL)** → comportamiento del usuario dentro de la plataforma  
- **experiment_checkout_ui.csv** → resultados de un experimento A/B en el checkout  

El análisis sigue una lógica clara y progresiva:

1. 🔍 Evaluar si podemos confiar en los datos (calidad de datos en Python)

2. 💰 Analizar si el negocio es rentable (revenue, costos y profit)  

3. 🛒 Entender dónde se pierden los usuarios (funnel de conversión)  

4. 🔁 Evaluar si los usuarios regresan (retención por cohortes)  

5. 🧪 Validar si los cambios generan impacto (test estadístico)  

6. 📊 Comunicar los resultados (dashboard en BI)  

A lo largo del proyecto, se transforman datos en insights para responder preguntas clave del negocio y proponer **recomendaciones accionables**.

---

## 🔹 Paso 1: Cargar y validar la calidad de los datos

---

### 1.1 Carga de datos y vista rápida

**🎯 Objetivo:** Familiarizarte con la estructura de los datasets del negocio antes de analizarlos.

**Instrucciones:**

- Importa las librerías necesarias
- Carga los archivos:
  - `rappiplus_orders_raw.csv`
  - `rappiplus_catalog.csv`
  - `rappiplus_marketing_spend.csv`
- Guarda los DataFrames en:
  - `orders`, `catalog`, `marketing`
- Explora cada dataset.

---

In [2]:
# 🔹 PASO 1 - CARGAR Y VALIDADR LA CALIDAD DE LOS DATOS
# importar librerías
import pandas as pd
import numpy as np

In [3]:
# cargar archivos
orders = pd.read_csv('https://practicum-content.s3.amazonaws.com/datasets/rappiplus_orders_raw.csv')
catalog = pd.read_csv('https://practicum-content.s3.amazonaws.com/datasets/rappiplus_catalog.csv')
marketing = pd.read_csv('https://practicum-content.s3.amazonaws.com/datasets/rappiplus_marketing_spend.csv')

In [4]:
# explorar datasets
# Explorar la estructura de orders
display("Orders crudo: vista general de la tabla", orders.head())
print(f"\n")
display("Orders crudo: información general del dataset", orders.info())
print(f"\n")
# Explorar columnas numéricas de orders
display("Orders crudo: estadísticas de columnas numéricas", orders.describe())
print(f"\n")
# Explorar columnas categóricas de orders
categoricas_orders = ['id_pedido', 'id_usuario', 'pais','dispositivo','fuente_referencia','nombre_producto','categoria_producto']
for col in categoricas_orders:
  print(f"\n")
  display("Orders crudo: estadísticas de columnas categóricas",orders[col].value_counts())
  print(f"\n")
# Explorar la estructura de catalog
display("Catalog crudo: vista general de la tabla", catalog.head())
print(f"\n")
display("Catalog crudo: información general del dataset", catalog.info())
print(f"\n")
# Explorar columnas numéricas de catalog
display("Catalog crudo: estadísticas de columnas numéricas", catalog.describe())
print(f"\n")
# Explorar columnas categóricas de catalog
categoricas_catalog = ['nombre_producto', 'categoria_producto', 'proveedor']
for col in categoricas_catalog:
  print(f"\n")
  display("Catalog crudo: estadísticas de columnas categóricas",catalog[col].value_counts())
  print(f"\n")
# Explorar la estructura de marketing
display("Marketing crudo: vista general de la tabla", marketing.head())
print(f"\n")
display("Marketing crudo: información general del dataset", marketing.info())
print(f"\n")
# Explorar columnas numéricas de marketing
display("Marketing crudo: estadísticas de columnas numéricas", marketing.describe())
print(f"\n")
# Explorar columnas categóricas de marketing
categoricas_marketing = ['pais', 'id_campaña', 'canal']
for col in categoricas_marketing:
  print(f"\n")
  display("Marketing crudo: estadísticas de columnas categóricas",marketing[col].value_counts())

'Orders crudo: vista general de la tabla'

,id_pedido,id_usuario,fecha_hora_pedido,pais,dispositivo,fuente_referencia,nombre_producto,categoria_producto,cantidad,precio_unitario,monto_descuento,monto_total
0,order_0,user_6993,2025-05-22,Argentina,desktop,organic,Jacket-Winter-M,Moda,2.0,332.69,0.0,665.37
1,order_1,user_1329,2025-06-15,Mexico,desktop,paid_search,Tablet-Standard-64GB,Electronica,1.0,176.86,5.0,171.86
2,order_2,user_3194,2025-05-02,Argentina,desktop,social,Blender-XL-Red,Hogar,2.0,102.99,10.0,195.99
3,order_3,user_4510,2025-06-09,Colombia,mobile,social,Tablet-Standard-64GB,Electronica,1.0,257.87,15.0,242.87
4,order_4,user_5044,2025-03-30,Argentina,desktop,paid_search,Blender-XL-Red,Hogar,1.0,336.28,0.0,336.28




<class 'pandas.core.frame.DataFrame'>
RangeIndex: 25100 entries, 0 to 25099
Data columns (total 12 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   id_pedido           25100 non-null  object 
 1   id_usuario          25100 non-null  object 
 2   fecha_hora_pedido   25100 non-null  object 
 3   pais                24800 non-null  object 
 4   dispositivo         25080 non-null  object 
 5   fuente_referencia   25070 non-null  object 
 6   nombre_producto     25070 non-null  object 
 7   categoria_producto  25020 non-null  object 
 8   cantidad            25050 non-null  float64
 9   precio_unitario     25050 non-null  float64
 10  monto_descuento     25050 non-null  float64
 11  monto_total         25100 non-null  float64
dtypes: float64(4), object(8)
memory usage: 2.3+ MB


'Orders crudo: información general del dataset'

None

'Orders crudo: estadísticas de columnas numéricas'

,cantidad,precio_unitario,monto_descuento,monto_total
count,25050.000000,25050.000000,25050.000000,2.510000e+04
mean,7.092735,259.305549,4.500798,2.072680e+03
std,296.277003,138.726461,5.223010,9.894995e+04
min,-2.000000,20.030000,0.000000,-4.926500e+02
25%,1.000000,138.377500,0.000000,1.805075e+02
50%,2.000000,258.715000,0.000000,3.417500e+02
75%,2.000000,380.332500,10.000000,5.185800e+02
max,20000.000000,499.960000,15.000000,8.840200e+06


'Orders crudo: estadísticas de columnas categóricas'

,count
id_pedido,
order_22936,2
order_7321,2
order_2470,2
order_7396,2
order_13811,2
...,...
order_8346,1
order_8345,1
order_8344,1


'Orders crudo: estadísticas de columnas categóricas'

,count
id_usuario,
user_7769,11
user_1822,10
user_3049,10
user_3000,10
user_5747,10
...,...
user_2415,1
user_2735,1
user_7205,1


'Orders crudo: estadísticas de columnas categóricas'

,count
pais,
Colombia,7520
Mexico,7502
Argentina,7291
mexico,865
colombia,823
argentina,799


'Orders crudo: estadísticas de columnas categóricas'

,count
dispositivo,
desktop,12759
mobile,12321


'Orders crudo: estadísticas de columnas categóricas'

,count
fuente_referencia,
social,8428
organic,8329
paid_search,8313


'Orders crudo: estadísticas de columnas categóricas'

,count
nombre_producto,
Vacuum-Pro-Black,4199
Blender-XL-Red,4195
Jacket-Winter-M,4192
Sneakers-Urban-42,4160
Laptop-Gaming-16GB,2794
Tablet-Standard-64GB,2780
Phone-Pro-128GB,2750


'Orders crudo: estadísticas de columnas categóricas'

,count
categoria_producto,
Hogar,8385
Moda,8323
Electronica,8312


'Catalog crudo: vista general de la tabla'

,nombre_producto,categoria_producto,costo_unitario,proveedor
0,Laptop-Gaming-16GB,Electrónica,280.68,"Fuller, Pena and Myers"
1,Phone-Pro-128GB,Electrónica,10.12,King Ltd
2,Tablet-Standard-64GB,Electrónica,25.21,Bowers LLC
3,Blender-XL-Red,Hogar,176.64,Long-Reid
4,Vacuum-Pro-Black,Hogar,16.60,"Rivera, Carr and Finley"




<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7 entries, 0 to 6
Data columns (total 4 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   nombre_producto     7 non-null      object 
 1   categoria_producto  7 non-null      object 
 2   costo_unitario      7 non-null      float64
 3   proveedor           7 non-null      object 
dtypes: float64(1), object(3)
memory usage: 356.0+ bytes


'Catalog crudo: información general del dataset'

None

'Catalog crudo: estadísticas de columnas numéricas'

,costo_unitario
count,7.000000
mean,102.252857
std,111.011563
min,10.120000
25%,16.905000
50%,25.210000
75%,182.975000
max,280.680000


'Catalog crudo: estadísticas de columnas categóricas'

,count
nombre_producto,
Laptop-Gaming-16GB,1
Phone-Pro-128GB,1
Tablet-Standard-64GB,1
Blender-XL-Red,1
Vacuum-Pro-Black,1
Sneakers-Urban-42,1
Jacket-Winter-M,1


'Catalog crudo: estadísticas de columnas categóricas'

,count
categoria_producto,
Electrónica,3
Hogar,2
Moda,2


'Catalog crudo: estadísticas de columnas categóricas'

,count
proveedor,
"Fuller, Pena and Myers",1
King Ltd,1
Bowers LLC,1
Long-Reid,1
"Rivera, Carr and Finley",1
Greene-Smith,1
Mcmillan-Rhodes,1


'Marketing crudo: vista general de la tabla'

,fecha,pais,id_campaña,canal,gasto
0,2025-01-01,Mexico,organic_Mexico,organic,2446.25
1,2025-01-01,Mexico,paid_search_Mexico,paid_search,2704.34
2,2025-01-01,Mexico,social_Mexico,social,2045.01
3,2025-01-01,Colombia,organic_Colombia,organic,2597.21
4,2025-01-01,Colombia,paid_search_Colombia,paid_search,1771.40




<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1620 entries, 0 to 1619
Data columns (total 5 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   fecha       1620 non-null   object 
 1   pais        1620 non-null   object 
 2   id_campaña  1620 non-null   object 
 3   canal       1519 non-null   object 
 4   gasto       1620 non-null   float64
dtypes: float64(1), object(4)
memory usage: 63.4+ KB


'Marketing crudo: información general del dataset'

None

'Marketing crudo: estadísticas de columnas numéricas'

,gasto
count,1620.00000
mean,1772.74292
std,734.43294
min,501.11000
25%,1128.03000
50%,1782.42500
75%,2420.68500
max,2999.36000


'Marketing crudo: estadísticas de columnas categóricas'

,count
pais,
Mexico,540
Colombia,540
Argentina,540


'Marketing crudo: estadísticas de columnas categóricas'

,count
id_campaña,
organic_Mexico,180
paid_search_Mexico,180
social_Mexico,180
organic_Colombia,180
paid_search_Colombia,180
social_Colombia,180
organic_Argentina,180
paid_search_Argentina,180
social_Argentina,180


'Marketing crudo: estadísticas de columnas categóricas'

,count
canal,
paid_search,507
organic,506
social,506


# 🔹 Paso 1: Preparación y calidad de datos

## 1.1 Carga de datos y exploración inicial

### 📊 Dataset: orders

**Estructura general:**
- 25,100 registros y 12 columnas
- Tipos de datos:
  - 8 columnas categóricas (object)
  - 4 numéricas (float64)
- La columna `fecha_hora_pedido` está en formato string (requiere conversión a datetime)

### ⚠️ Hallazgos de calidad

#### 1. Valores nulos
Se identifican valores faltantes en varias columnas:

| Columna              | Valores nulos | % aproximado |
|---------------------|--------------|--------------|
| pais                | 300          | ~1.2%        |
| dispositivo         | 20           | ~0.08%       |
| fuente_referencia   | 30           | ~0.12%       |
| nombre_producto     | 30           | ~0.12%       |
| categoria_producto  | 80           | ~0.32%       |
| cantidad            | 50           | ~0.20%       |
| precio_unitario     | 50           | ~0.20%       |
| monto_descuento     | 50           | ~0.20%       |

👉 Esto puede afectar cálculos de revenue, segmentación y análisis de comportamiento.

---

#### 2. Duplicados en identificadores
- `id_pedido` debería ser único, pero existen duplicados (ej: algunos pedidos aparecen 2 veces)

👉 Riesgo de **sobreestimación de ingresos**.

---

#### 3. Inconsistencias en variables categóricas
- `pais` presenta duplicados semánticos por formato:
  - "Mexico" vs "mexico"
  - "Colombia" vs "colombia"
  - "Argentina" vs "argentina"

👉 Problema de **estandarización** que afecta agregaciones.

---

#### 4. Valores atípicos y errores en variables numéricas
- `cantidad`:
  - Valores negativos (-2) ❌
  - Valores extremadamente altos (hasta 20,000) ⚠️
- `monto_total`:
  - Valores negativos ❌
  - Máximos extremadamente altos (~8.8M) ⚠️

👉 Posibles errores de captura o registros corruptos.

---

#### 5. Inconsistencias en montos
- `monto_total` debería cumplir:
  - monto_total = cantidad * precio_unitario - monto_descuento
- Presencia de valores negativos o incoherentes indica inconsistencias

---

#### 6. Distribución de variables categóricas
- Canales (`fuente_referencia`) balanceados: organic, social, paid_search
- Productos y categorías bien distribuidos

👉 No hay sesgos evidentes en distribución, pero sí problemas de calidad.

---

---

### 📊 Dataset: catalog

**Estructura general:**
- 7 registros (catálogo pequeño)
- Sin valores nulos
- Tipos de datos correctos

### ⚠️ Hallazgos

#### 1. Inconsistencia en categorías
- Aparece "Electrónica" (con tilde), mientras que en orders es "Electronica"

👉 Problema de **normalización de texto** que afectará joins.

---

#### 2. Integridad general
- No hay duplicados
- Cada producto es único
- Costos dentro de rangos razonables

👉 Dataset limpio, pero requiere estandarización de texto.

---

---

### 📊 Dataset: marketing

**Estructura general:**
- 1,620 registros
- 5 columnas
- `fecha` en formato string (requiere conversión)

### ⚠️ Hallazgos

#### 1. Valores nulos

| Columna | Valores nulos | % aproximado |
|--------|--------------|--------------|
| canal  | 101          | ~6.2%        |

---

#### 2. Oportunidad de imputación (hallazgo clave)
- La columna `id_campaña` contiene información del canal:
- Ejemplo: `paid_search_Mexico`

👉 Se puede:
- Extraer el canal (`paid_search`)
- Eliminar el sufijo del país (`_Mexico`)
- Completar los valores nulos en `canal`

✔️ Esto evita clasificar datos como "unknown" y mejora la calidad del dataset.

---

#### 3. Consistencia categórica
- Países bien definidos (sin duplicados por formato)
- Canales esperados: organic, social, paid_search

---

#### 4. Variables numéricas
- `gasto` sin valores negativos
- Distribución dentro de rangos razonables

👉 No presenta problemas críticos.

---

## 🧩 Conclusión general

| DataFrame               | Estado general | Principales acciones |
|------------------------|---------------|---------------------|
| orders             | Malo         | Limpiar nulos, convertir fechas, filtrar o corregir valores negativos, evaluar tratamiento de outliers, corregir texto columnas categoricas|
| catalog                | Bueno     | Corregir en la columna categoria_producto "Electrónica" (eliminar tilde)|
| marketing        | Malo     | Limpiar nulos de la columna canal, convertir fechas |

## 1.2 Revisión y calidad de datos

### 🎯 Objetivo
Asegurar que los datos sean confiables para el análisis de:
- Revenue
- Costos
- Rentabilidad

---

### 🔧 Validaciones necesarias

#### 1. Conversión de fechas
- Convertir:
- `fecha_hora_pedido` (orders) → datetime
- `fecha` (marketing) → datetime

👉 Necesario para análisis temporal.

---

#### 2. Revisión de variables numéricas
- Eliminar o tratar:
- Cantidades negativas
- Montos negativos
- Valores extremos (outliers)

👉 Evita distorsiones en KPIs.

---

#### 3. Validación de consistencia de montos
- Verificar que:
   - monto_total ≈ cantidad * precio_unitario - monto_descuento

👉 Detectar y corregir inconsistencias.

---

#### 4. Manejo de duplicados
- Revisar duplicados en:
- `id_pedido` (orders)

👉 Eliminar duplicados para evitar doble conteo.

---

#### 5. Tratamiento de valores nulos
- Estrategias:
- Eliminar registros críticos si faltan variables clave
- Imputar cuando sea posible

📌 Caso importante:
- En `marketing.canal`, completar nulos usando `id_campaña`

---

#### 6. Normalización de variables categóricas
- Estandarizar:
- Países (minúsculas/mayúsculas)
- Categorías ("Electrónica" vs "Electronica")

👉 Clave para joins y agregaciones correctas.

---

### ✅ Conclusión del Paso 1

Los datasets presentan múltiples problemas de calidad que deben ser corregidos antes del análisis:

- Nulos en variables clave  
- Duplicados en identificadores  
- Inconsistencias en texto  
- Valores inválidos en métricas numéricas  
- Oportunidades de imputación inteligente (marketing)

👉 La limpieza adecuada permitirá construir métricas confiables en los siguientes pasos del proyecto.


In [5]:
# Conversión de fechas

orders['fecha_hora_pedido'] = pd.to_datetime(orders['fecha_hora_pedido'], errors='coerce')
marketing['fecha'] = pd.to_datetime(marketing['fecha'], errors='coerce')

In [6]:
# Limpieza dataset Orders

# Estandarizar texto
orders['pais'] = orders['pais'].str.strip().str.lower().str.capitalize()

# Eliminar duplicados por id_pedido
orders = orders.drop_duplicates(subset=['id_pedido'])

# Eliminar valores numéricos inválidos
orders = orders[
    (orders['cantidad'] > 0) &
    (orders['precio_unitario'] > 0) &
    (orders['monto_total'] > 0)
]

# Eliminar outliers extremos
orders = orders[orders['cantidad'] < 1000]

# Validar consistencia de montos
orders['monto_calculado'] = orders['cantidad'] * orders['precio_unitario'] - orders['monto_descuento']

# Mantener registros consistentes (tolerancia pequeña por decimales)
orders = orders[np.isclose(orders['monto_total'], orders['monto_calculado'], atol=1)]

# Eliminar columna auxiliar
orders = orders.drop(columns=['monto_calculado'])

# Manejo de nulos
orders = orders.dropna(subset=[
    'pais', 'dispositivo', 'fuente_referencia',
    'nombre_producto', 'categoria_producto'
])


In [7]:
# Limpieza dataset Catalog

# Estandarizar categorías (quitar tilde)
catalog['categoria_producto'] = catalog['categoria_producto'].str.replace('Electrónica', 'Electronica')

In [8]:
# Limpieza dataset Marketing

# Completar nulos en canal usando id_campaña
marketing['canal'] = marketing['canal'].fillna(
    marketing['id_campaña'].str.rsplit('_', n=1).str[0]
)

# Estandarizar texto en país
marketing['pais'] = marketing['pais'].str.strip().str.lower().str.capitalize()


In [9]:
# Validación final

# Explorar la estructura de orders
display("Orders limpio: vista general de la tabla", orders.head())
print(f"\n")
display("Orders limpio: información general del dataset", orders.info())
print(f"\n")
# Explorar columnas numéricas de orders
display("Orders limpio: estadísticas de columnas numéricas", orders.describe())
print(f"\n")
# Explorar columnas categóricas de orders
categoricas_orders = ['id_pedido', 'id_usuario', 'pais','dispositivo','fuente_referencia','nombre_producto','categoria_producto']
for col in categoricas_orders:
  print(f"\n")
  display("Orders limpio: estadísticas de columnas categóricas",orders[col].value_counts())
  print(f"\n")
# Explorar la estructura de catalog
display("Catalog limpio: vista general de la tabla", catalog.head())
print(f"\n")
display("Catalog limpio: información general del dataset", catalog.info())
print(f"\n")
# Explorar columnas numéricas de catalog
display("Catalog limpio: estadísticas de columnas numéricas", catalog.describe())
print(f"\n")
# Explorar columnas categóricas de catalog
categoricas_catalog = ['nombre_producto', 'categoria_producto', 'proveedor']
for col in categoricas_catalog:
  print(f"\n")
  display("Catalog limpio: estadísticas de columnas categóricas",catalog[col].value_counts())
  print(f"\n")
# Explorar la estructura de marketing
display("Marketing limpio: vista general de la tabla", marketing.head())
print(f"\n")
display("Marketing limpio: información general del dataset", marketing.info())
print(f"\n")
# Explorar columnas numéricas de marketing
display("Marketing limpio: estadísticas de columnas numéricas", marketing.describe())
print(f"\n")
# Explorar columnas categóricas de marketing
categoricas_marketing = ['pais', 'id_campaña', 'canal']
for col in categoricas_marketing:
  print(f"\n")
  display("Marketing limpio: estadísticas de columnas categóricas",marketing[col].value_counts())

'Orders limpio: vista general de la tabla'

,id_pedido,id_usuario,fecha_hora_pedido,pais,dispositivo,fuente_referencia,nombre_producto,categoria_producto,cantidad,precio_unitario,monto_descuento,monto_total
0,order_0,user_6993,2025-05-22,Argentina,desktop,organic,Jacket-Winter-M,Moda,2.0,332.69,0.0,665.37
1,order_1,user_1329,2025-06-15,Mexico,desktop,paid_search,Tablet-Standard-64GB,Electronica,1.0,176.86,5.0,171.86
2,order_2,user_3194,2025-05-02,Argentina,desktop,social,Blender-XL-Red,Hogar,2.0,102.99,10.0,195.99
3,order_3,user_4510,2025-06-09,Colombia,mobile,social,Tablet-Standard-64GB,Electronica,1.0,257.87,15.0,242.87
4,order_4,user_5044,2025-03-30,Argentina,desktop,paid_search,Blender-XL-Red,Hogar,1.0,336.28,0.0,336.28




<class 'pandas.core.frame.DataFrame'>
Index: 24590 entries, 0 to 24999
Data columns (total 12 columns):
 #   Column              Non-Null Count  Dtype         
---  ------              --------------  -----         
 0   id_pedido           24590 non-null  object        
 1   id_usuario          24590 non-null  object        
 2   fecha_hora_pedido   24590 non-null  datetime64[ns]
 3   pais                24590 non-null  object        
 4   dispositivo         24590 non-null  object        
 5   fuente_referencia   24590 non-null  object        
 6   nombre_producto     24590 non-null  object        
 7   categoria_producto  24590 non-null  object        
 8   cantidad            24590 non-null  float64       
 9   precio_unitario     24590 non-null  float64       
 10  monto_descuento     24590 non-null  float64       
 11  monto_total         24590 non-null  float64       
dtypes: datetime64[ns](1), float64(4), object(7)
memory usage: 2.4+ MB


'Orders limpio: información general del dataset'

None

'Orders limpio: estadísticas de columnas numéricas'

,fecha_hora_pedido,cantidad,precio_unitario,monto_descuento,monto_total
count,24590,24590.000000,24590.000000,24590.000000,24590.000000
mean,2025-04-01 03:23:19.300528640,1.505205,259.378319,4.504677,385.997362
min,2025-01-01 00:00:00,1.000000,20.030000,0.000000,5.240000
25%,2025-02-15 00:00:00,1.000000,138.737500,0.000000,180.615000
50%,2025-04-02 00:00:00,2.000000,258.760000,0.000000,341.540000
75%,2025-05-16 00:00:00,2.000000,380.377500,10.000000,518.640000
max,2025-06-30 00:00:00,2.000000,499.960000,15.000000,999.890000
std,NaN,0.499983,138.688181,5.225474,255.800606


'Orders limpio: estadísticas de columnas categóricas'

,count
id_pedido,
order_24999,1
order_0,1
order_1,1
order_2,1
order_3,1
...,...
order_18,1
order_17,1
order_16,1


'Orders limpio: estadísticas de columnas categóricas'

,count
id_usuario,
user_7769,11
user_3000,10
user_4065,10
user_1822,10
user_5111,10
...,...
user_1457,1
user_2816,1
user_6163,1


'Orders limpio: estadísticas de columnas categóricas'

,count
pais,
Mexico,8300
Colombia,8271
Argentina,8019


'Orders limpio: estadísticas de columnas categóricas'

,count
dispositivo,
desktop,12519
mobile,12071


'Orders limpio: estadísticas de columnas categóricas'

,count
fuente_referencia,
social,8265
organic,8188
paid_search,8137


'Orders limpio: estadísticas de columnas categóricas'

,count
nombre_producto,
Jacket-Winter-M,4122
Blender-XL-Red,4114
Vacuum-Pro-Black,4114
Sneakers-Urban-42,4072
Laptop-Gaming-16GB,2742
Tablet-Standard-64GB,2729
Phone-Pro-128GB,2697


'Orders limpio: estadísticas de columnas categóricas'

,count
categoria_producto,
Hogar,8228
Moda,8194
Electronica,8168


'Catalog limpio: vista general de la tabla'

,nombre_producto,categoria_producto,costo_unitario,proveedor
0,Laptop-Gaming-16GB,Electronica,280.68,"Fuller, Pena and Myers"
1,Phone-Pro-128GB,Electronica,10.12,King Ltd
2,Tablet-Standard-64GB,Electronica,25.21,Bowers LLC
3,Blender-XL-Red,Hogar,176.64,Long-Reid
4,Vacuum-Pro-Black,Hogar,16.60,"Rivera, Carr and Finley"




<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7 entries, 0 to 6
Data columns (total 4 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   nombre_producto     7 non-null      object 
 1   categoria_producto  7 non-null      object 
 2   costo_unitario      7 non-null      float64
 3   proveedor           7 non-null      object 
dtypes: float64(1), object(3)
memory usage: 356.0+ bytes


'Catalog limpio: información general del dataset'

None

'Catalog limpio: estadísticas de columnas numéricas'

,costo_unitario
count,7.000000
mean,102.252857
std,111.011563
min,10.120000
25%,16.905000
50%,25.210000
75%,182.975000
max,280.680000


'Catalog limpio: estadísticas de columnas categóricas'

,count
nombre_producto,
Laptop-Gaming-16GB,1
Phone-Pro-128GB,1
Tablet-Standard-64GB,1
Blender-XL-Red,1
Vacuum-Pro-Black,1
Sneakers-Urban-42,1
Jacket-Winter-M,1


'Catalog limpio: estadísticas de columnas categóricas'

,count
categoria_producto,
Electronica,3
Hogar,2
Moda,2


'Catalog limpio: estadísticas de columnas categóricas'

,count
proveedor,
"Fuller, Pena and Myers",1
King Ltd,1
Bowers LLC,1
Long-Reid,1
"Rivera, Carr and Finley",1
Greene-Smith,1
Mcmillan-Rhodes,1


'Marketing limpio: vista general de la tabla'

,fecha,pais,id_campaña,canal,gasto
0,2025-01-01,Mexico,organic_Mexico,organic,2446.25
1,2025-01-01,Mexico,paid_search_Mexico,paid_search,2704.34
2,2025-01-01,Mexico,social_Mexico,social,2045.01
3,2025-01-01,Colombia,organic_Colombia,organic,2597.21
4,2025-01-01,Colombia,paid_search_Colombia,paid_search,1771.40




<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1620 entries, 0 to 1619
Data columns (total 5 columns):
 #   Column      Non-Null Count  Dtype         
---  ------      --------------  -----         
 0   fecha       1620 non-null   datetime64[ns]
 1   pais        1620 non-null   object        
 2   id_campaña  1620 non-null   object        
 3   canal       1620 non-null   object        
 4   gasto       1620 non-null   float64       
dtypes: datetime64[ns](1), float64(1), object(3)
memory usage: 63.4+ KB


'Marketing limpio: información general del dataset'

None

'Marketing limpio: estadísticas de columnas numéricas'

,fecha,gasto
count,1620,1620.00000
mean,2025-03-31 12:00:00,1772.74292
min,2025-01-01 00:00:00,501.11000
25%,2025-02-14 18:00:00,1128.03000
50%,2025-03-31 12:00:00,1782.42500
75%,2025-05-15 06:00:00,2420.68500
max,2025-06-29 00:00:00,2999.36000
std,NaN,734.43294


'Marketing limpio: estadísticas de columnas categóricas'

,count
pais,
Mexico,540
Colombia,540
Argentina,540


'Marketing limpio: estadísticas de columnas categóricas'

,count
id_campaña,
organic_Mexico,180
paid_search_Mexico,180
social_Mexico,180
organic_Colombia,180
paid_search_Colombia,180
social_Colombia,180
organic_Argentina,180
paid_search_Argentina,180
social_Argentina,180


'Marketing limpio: estadísticas de columnas categóricas'

,count
canal,
organic,540
paid_search,540
social,540


---
**📦 Exportación**: Una vez finalizada la limpieza, se exportan los datasets para utilizarlos en la última etapa del proyecto.

In [10]:
# exportar datasets
orders.to_csv('orders_clean.csv', index=False)
catalog.to_csv('catalog_clean.csv', index=False)
marketing.to_csv('marketing_clean.csv', index=False)

---

# 🔹 Paso 2: Análisis de rentabilidad del negocio

## 2.1 Cálculo de KPIs principales

### 🎯 Objetivo: Calcular los indicadores clave del negocio para evaluar ingresos, costos y rentabilidad.

Se usan los 3 datasets (`orders_clean`, `catalog_clean`, `marketing_clean`):

---

# 📊 Parte 1: Rentabilidad del negocio

## 💰 ¿Cuál es el ingreso total (Revenue)?

👉 **Revenue total: 9,491,675.14**

Esto representa el total de ingresos generados por las ventas en la plataforma.

---

## 🏷️ ¿Cuál es el costo total?

👉 **Costo total: 3,783,478.80**

Corresponde al costo de los productos vendidos.

---

## 📢 ¿Cuánto se ha invertido en marketing?  

👉 **Gasto en marketing: 2,871,843.53**

Este valor representa la inversión total realizada en adquisición de usuarios.

---

## 📈 ¿El negocio es rentable? (Profit)

👉 **Profit: 2,836,352.81**

### 🧠 Interpretación:
- El negocio genera una utilidad positiva significativa  
- Los ingresos cubren tanto los costos como la inversión en marketing  

✅ **Conclusión:** El negocio es rentable  

---

# 📈 Parte 2: Comportamiento de ventas

## 🧾 ¿Cuál es el ticket promedio por orden?

👉 **Ticket promedio: 386.00**

### 🧠 Interpretación:
- En promedio, cada usuario gasta 386 por pedido  
- Indica un nivel de gasto relativamente alto por transacción  

---

## 📦 ¿Cuál es la cantidad promedio de productos por orden?

👉 **Cantidad promedio: 1.51 productos por orden**

### 🧠 Interpretación:
- La mayoría de pedidos contienen entre 1 y 2 productos  
- Existe oportunidad de aumentar el tamaño del carrito (cross-selling o bundles)

---

## 🏆 ¿Cuál es el producto más vendido?

👉 **Producto más vendido: Vacuum-Pro-Black**

### 🧠 Interpretación:
- Es el producto con mayor volumen de ventas  
- Puede ser un producto clave para:
  - Promociones
  - Estrategias de marketing
  - Optimización de inventario  

---

## 📢 ¿Cuánto se ha gastado en marketing por canal?

| Canal         | Gasto        |
|--------------|-------------|
| social        | 976,818.37  |
| organic       | 972,650.96  |
| paid_search   | 922,374.20  |

### 🧠 Interpretación:
- La inversión está relativamente balanceada entre los tres canales  
- El canal **social** tiene la mayor inversión  
- No hay una dependencia excesiva de un solo canal  

---

# 📌 Insights clave del negocio

- El negocio es **altamente rentable** con más de 2.8M en profit  
- El ticket promedio es alto, lo cual es positivo para ingresos  
- El volumen por orden es bajo → oportunidad de aumentar unidades por compra  
- La inversión en marketing está equilibrada → buena diversificación  

---

# ✅ Conclusión del Paso 2

- RappiPlus es un negocio rentable y sostenible  
- Existen oportunidades claras de optimización en:
  - Incremento del ticket (más productos por orden)
  - Estrategias sobre productos estrella
  - Optimización del gasto en marketing  

👉 Este análisis sienta las bases para decisiones estratégicas en crecimiento y rentabilidad.

In [11]:
# 🔹 PASO 2 - ANALIZAR SI EL NEGOCIO ES RENTABLE

# Cargar datasets limpios

orders = pd.read_csv('/content/orders_clean.csv')
catalog = pd.read_csv('/content/catalog_clean.csv')
marketing = pd.read_csv('/content/marketing_clean.csv')

# Merge para traer costo_unitario al dataset de orders
orders_full = orders.merge(
    catalog[['nombre_producto', 'costo_unitario']],
    on='nombre_producto',
    how='left'
)

# ------------------------------
# PARTE 1: RENTABILIDAD
# ------------------------------

# Revenue total
revenue = orders_full['monto_total'].sum()

# Costo total
orders_full['costo_total'] = orders_full['cantidad'] * orders_full['costo_unitario']
costo_total = orders_full['costo_total'].sum()

# Gasto en marketing
marketing_total = marketing['gasto'].sum()

# 2.4 Profit
profit = revenue - costo_total - marketing_total

# ------------------------------
# PARTE 2: COMPORTAMIENTO
# ------------------------------

# Número de órdenes
num_orders = orders_full['id_pedido'].nunique()

# Ticket promedio
ticket_promedio = revenue / num_orders

# Cantidad promedio de productos por orden
cantidad_promedio = orders_full['cantidad'].sum() / num_orders

# Producto más vendido
producto_mas_vendido = (
    orders_full.groupby('nombre_producto')['cantidad']
    .sum()
    .sort_values(ascending=False)
    .idxmax()
)

# Gasto en marketing por canal
gasto_por_canal = marketing.groupby('canal')['gasto'].sum().sort_values(ascending=False)

# ------------------------------
# 3. RESULTADOS
# ------------------------------

print("--- RENTABILIDAD ---")
print(f"Revenue total: {revenue:,.2f}")
print(f"Costo total: {costo_total:,.2f}")
print(f"Gasto en marketing: {marketing_total:,.2f}")
print(f"Profit: {profit:,.2f}")

if profit > 0:
    print("El negocio es rentable")
else:
    print("El negocio NO es rentable")

print("\n --- COMPORTAMIENTO DE VENTAS ---")
print(f"Ticket promedio: {ticket_promedio:,.2f}")
print(f"Cantidad promedio por orden: {cantidad_promedio:,.2f}")
print(f"Producto más vendido: {producto_mas_vendido}")

print("\n --- GASTO POR CANAL ---")
display(gasto_por_canal)


--- RENTABILIDAD ---
Revenue total: 9,491,675.14
Costo total: 3,783,478.80
Gasto en marketing: 2,871,843.53
Profit: 2,836,352.81
El negocio es rentable

 --- COMPORTAMIENTO DE VENTAS ---
Ticket promedio: 386.00
Cantidad promedio por orden: 1.51
Producto más vendido: Vacuum-Pro-Black

 --- GASTO POR CANAL ---


,gasto
canal,
social,976818.37
organic,972650.96
paid_search,922374.20


## 🔹 Paso 3: Entender dónde se pierden los usuarios (funnel de conversión)

**🎯 Objetivo:** Analizar el comportamiento de los usuarios para identificar en qué etapa del proceso se pierden.


⚙️**Conexión a la base de datos**:  
Se ejecuta la línea de configuración para conectar con la base de datos y aplicar consultas SQL en la tabla **events**.

---

## 📊 Parte 1: Construcción del funnel

### 🧩 ¿Cuántos usuarios llegan a cada etapa del funnel?

A partir de la consulta SQL, se obtuvo el número de usuarios únicos por evento:

| Etapa               | Usuarios |
|--------------------|----------|
| first_visit        | 7,796    |
| add_to_cart        | 7,634    |
| select_item        | 7,582    |
| begin_checkout     | 7,208    |
| add_payment_info   | 6,250    |
| purchase           | 6,240    |

👉 Estos valores representan la cantidad de usuarios que alcanzan cada etapa del funnel.

---

### 🔄 Orden del funnel (flujo del usuario)

Con base en los eventos observados en los datos, el funnel real es:

1. `first_visit` → Usuario llega por primera vez  
2. `add_to_cart` → Usuario selecciona y agrega producto al carrito
3. `select_item` → Usuario selecciona producto dentro del carrito para iniciar la compra
4. `begin_checkout` → Usuario inicia el proceso de compra  
5. `add_payment_info`→ Usuario ingresa información de pago    
6. `purchase` → Usuario completa la compra  
---

### 📊 Análisis inicial del funnel

- El mayor volumen de usuarios entra en `first_visit` (7,796 usuarios)
- Existe una leve caída entre:
  - `first_visit` → `add_to_cart`
  - `add_to_cart` → `select_item`
  - `select_item` → `begin_checkout  `
- La caída más relevante comienza en:
  - `begin_checkout` → `add_payment_info`
- La diferencia entre `add_payment_info` y `purchase` es mínima (solo 10 usuarios)

---

### 🧠 Insight preliminar

👉 El funnel muestra un comportamiento saludable en las primeras etapas (alta retención), pero:

- Existe fricción importante en el proceso de pago (`add_payment_info`)
- El paso final (`purchase`) tiene una conversión casi total, lo cual indica que:
  - El problema no es cerrar la compra
  - El problema está antes, en el proceso de pago

---

### ✅ Conclusión Parte 1

- El funnel está bien estructurado y presenta alta retención inicial  
- El mayor cuello de botella se encuentra en la etapa de pago  
- Optimizar el proceso de checkout puede generar un impacto directo en ingresos  

👉 Este análisis prepara el camino para calcular las tasas de conversión en la siguiente sección.

---

## 📉 Parte 2: Análisis de conversión

### 📊 Tasas de conversión por etapa

A partir de la consulta SQL, se obtienen las siguientes métricas del funnel:

| Etapa               | Usuarios | Usuarios previos | Tasa conversión | Drop-off |
|--------------------|----------|------------------|------------------|----------|
| first_visit        | 7,796    | -                | -                | -        |
| add_to_cart        | 7,634    | 7,796            | 97%              | 3%       |
| select_item        | 7,582    | 7,634            | 99%              | 1%       |
| begin_checkout     | 7,208    | 7,582            | 95%              | 5%       |
| add_payment_info   | 6,250    | 7,208            | 86%              | 14%      |
| purchase           | 6,240    | 6,250            | 99%              | 1%       |

---

### 🔍 ¿Cómo se interpreta la tasa de conversión?

- La tasa de conversión indica qué porcentaje de usuarios avanza al siguiente paso
- El drop-off representa la pérdida de usuarios en cada etapa

👉 Idealmente:
- Conversión alta ✅
- Drop-off bajo ✅

---

### 🚨 ¿En qué etapa se pierden más usuarios?

La mayor caída ocurre en:

👉 **`add_payment_info` (14% de drop-off)**

Esto indica que:
- Es el principal cuello de botella del funnel
- Los usuarios abandonan justo antes de completar la compra

---

### 🧠 Interpretación del comportamiento

#### 🔹 Etapas iniciales (muy eficientes)
- `first_visit → add_to_cart` → 97%
- `add_to_cart → select_item` → 99%

👉 Indican:
- Buena experiencia de navegación
- Interés en los productos

---

#### 🔹 Etapa media (ligera fricción)
- `select_item → begin_checkout` → 95%

👉 Posibles causas:
- Costos inesperados
- Falta de urgencia para comprar

---

#### 🔹 Etapa crítica (problema principal)
- `begin_checkout → add_payment_info` → 86%

👉 Posibles causas:
- Proceso de pago complejo
- Falta de confianza
- Métodos de pago limitados
- Costos adicionales visibles en checkout

---

#### 🔹 Etapa final (muy eficiente)
- `add_payment_info → purchase` → 99%

👉 Indica:
- Una vez el usuario ingresa datos de pago, casi siempre compra
- El problema NO está en el cierre

---

### 🎯 ¿Cuál es la tasa de conversión final?

👉 **Conversión final: 80.04%**

Esto significa que:
- De cada 100 usuarios que entran al funnel, aproximadamente 80 completan una compra

---

### 📌 Insight clave de negocio

- El funnel es altamente eficiente en general (80% es una conversión muy alta)
- Sin embargo, existe una oportunidad clara de mejora en el checkout

👉 **Optimizar la etapa de pago puede generar un aumento directo en revenue sin aumentar el tráfico**

---

### ✅ Conclusión Parte 2

- El funnel tiene un desempeño sólido en adquisición y navegación  
- El principal problema está en la etapa de pago  
- Mejorar el checkout es la acción de mayor impacto  

👉 Este análisis permite priorizar mejoras estratégicas enfocadas en conversión.

In [12]:
import pandas as pd
from sqlalchemy import create_engine

# ======================
# Conexión (NO modificar)
# ======================
db_config = {
    'user': 'practicum_student',
    'pwd': 'QnmDH8Sc2TQLvy2G3Vvh7',
    'host': 'yp-trainers-practicum.cluster-czs0gxyx2d8w.us-east-1.rds.amazonaws.com',
    'port': 5432,
    'db': 'data-analyst-production-db-en'
}

connection_string = 'postgresql://{}:{}@{}:{}/{}'.format(
    db_config['user'],
    db_config['pwd'],
    db_config['host'],
    db_config['port'],
    db_config['db']
)

engine = create_engine(connection_string, connect_args={'sslmode':'require'})

In [13]:
# 🔹 PASO 3 - ENTENDER DÓNDE SE PIERDEN LOS USUARIOS (FUNNEL DE CONVERSIÓN)
# Explorar tabla events
# =========================
query_events = '''
SELECT *
FROM events
LIMIT 10;
'''
events = pd.read_sql(query_events, con=engine)
events.head(10)

,id_usuario,id_sesion,nombre_evento,timestamp_evento,pais,dispositivo,fuente_referencia,categoria_producto
0,user_6772,6a97f2af-32ae-4186-8c92-04025be1a27b,first_visit,2025-05-17,Colombia,desktop,organic,Moda
1,user_5883,369b767c-1c33-4b2f-a652-c7c0ef92cfc9,add_to_cart,2025-02-23,Mexico,mobile,social,Hogar
2,user_5946,60039041-e78b-474c-87b3-c0b7e9c30708,add_payment_info,2025-05-15,Colombia,desktop,social,Electronica
3,user_827,18252a64-f389-4ef7-9e58-dadad4a3491e,purchase,2025-03-31,Mexico,mobile,social,Moda
4,user_2361,221b364e-cdc5-4668-b698-18d5ba849a67,first_visit,2025-01-22,Argentina,desktop,paid_search,Electronica
5,user_4402,0892643c-f80d-4eb4-9ca8-3638b16c6ed4,begin_checkout,2025-04-21,Argentina,desktop,paid_search,Hogar
6,user_767,0005a13c-3c1f-496b-a26a-f3c282f19e13,begin_checkout,2025-03-01,Mexico,mobile,social,Electronica
7,user_2694,b07336e6-1207-4ee1-82e3-1fa7fbd194e3,add_payment_info,2025-03-22,Colombia,desktop,organic,Hogar
8,user_7209,f42bd8cd-aff2-4b7e-94e2-d4d885d23afc,add_to_cart,2025-05-07,Mexico,desktop,organic,Moda
9,user_1788,abf1af77-e96b-4490-9efb-3fa5d8271ad4,add_to_cart,2025-05-13,Argentina,desktop,organic,Moda


In [14]:
# PARTE 1: Totales del funnel
# ======================

query_totals = '''
SELECT
    nombre_evento,
    COUNT(DISTINCT id_usuario) AS usuarios_unicos
FROM events
GROUP BY nombre_evento
ORDER BY usuarios_unicos DESC;
'''

totals = pd.read_sql(query_totals, con=engine)
totals

,nombre_evento,usuarios_unicos
0,first_visit,7796
1,add_to_cart,7634
2,select_item,7582
3,begin_checkout,7208
4,add_payment_info,6250
5,purchase,6240


In [15]:
# PARTE 2: Conversiones
# ======================

query_conversion = '''
-- Conteo de usuarios únicos por etapa del funnel y ordenar eventos según el flujo del usuario
WITH funnel AS (
    SELECT
        nombre_evento,
        COUNT(DISTINCT id_usuario) AS usuarios,
        CASE
            WHEN nombre_evento = 'first_visit' THEN 1
            WHEN nombre_evento = 'add_to_cart' THEN 2
            WHEN nombre_evento = 'select_item' THEN 3
            WHEN nombre_evento = 'begin_checkout' THEN 4
            WHEN nombre_evento = 'add_payment_info' THEN 5
            WHEN nombre_evento = 'purchase' THEN 6
        END AS orden
    FROM events
    GROUP BY nombre_evento
),
-- Calcular conversiones entre etapas
conversiones AS (
    SELECT
        nombre_evento,
        usuarios,
        orden,
        LAG(usuarios) OVER (ORDER BY orden) AS usuarios_previos,
        ROUND(usuarios * 100 / LAG(usuarios) OVER (ORDER BY orden), 2) AS tasa_conversion
    FROM funnel
),
--calculo de la tasa de conversion final
conversion_final AS (
    SELECT
        ROUND(
            (SELECT COUNT(DISTINCT id_usuario)
             FROM events
             WHERE nombre_evento = 'purchase') * 100.0
            /
            (SELECT COUNT(DISTINCT id_usuario)
             FROM events
             WHERE nombre_evento = 'first_visit'),
        2) AS conversion_final
)
-- resultado final del funnel
SELECT
    c.nombre_evento,
    c.usuarios,
    c.usuarios_previos,
    c.tasa_conversion,
    ROUND(100 - c.tasa_conversion, 2) AS drop_off,
    cf.conversion_final
FROM conversiones c
CROSS JOIN conversion_final cf
ORDER BY c.orden;
'''

conversion = pd.read_sql(query_conversion, con=engine)
conversion

,nombre_evento,usuarios,usuarios_previos,tasa_conversion,drop_off,conversion_final
0,first_visit,7796,NaN,NaN,NaN,80.04
1,add_to_cart,7634,7796.0,97.0,3.0,80.04
2,select_item,7582,7634.0,99.0,1.0,80.04
3,begin_checkout,7208,7582.0,95.0,5.0,80.04
4,add_payment_info,6250,7208.0,86.0,14.0,80.04
5,purchase,6240,6250.0,99.0,1.0,80.04


---

## 🔹 Paso 4: Evaluar si los usuarios regresan (retención por cohortes)

**🎯 Objetivo:** Analizar la retención de usuarios para entender si regresan después de registrarse.

**Tablas**

- `users`
- `user_activity`

---

# 📊 Construcción de cohortes

- Se agruparon los usuarios por mes de registro (`cohorte`)
- Se calculó la actividad en semanas posteriores:
  - Semana 1 (W1): días 1–7
  - Semana 2 (W2): días 8–14
  - Semana 3 (W3): días 15–21

---

# 📈 Resultados de retención

| Cohorte  | Usuarios iniciales | Retenido W1 | Retenido W2 | Retenido W3 | Semana 1 | Semana 2 | Semana 3 |
|----------|------------------|-------------|-------------|-------------|----------|----------|----------|
| 2025-01  | 1,627            | 697         | 668         | 656         | 42.84%   | 41.06%   | 40.32%   |
| 2025-02  | 1,444            | 611         | 609         | 635         | 42.31%   | 42.17%   | 43.98%   |
| 2025-03  | 1,636            | 677         | 705         | 690         | 41.38%   | 43.09%   | 42.18%   |
| 2025-04  | 1,606            | 680         | 697         | 663         | 42.34%   | 43.40%   | 41.28%   |
| 2025-05  | 1,687            | 695         | 676         | 706         | 41.20%   | 40.07%   | 41.85%   |

---

# 🔍 Análisis de retención

## 📌 Nivel general de retención

- La retención en semana 1 se mantiene estable alrededor del **41% - 43%**
- No hay caídas drásticas entre semanas
- La retención se mantiene relativamente constante en el tiempo

👉 Esto indica una **buena estabilidad en el engagement inicial**

---

## 📉 ¿Los usuarios regresan?

👉 Sí, aproximadamente:

- **4 de cada 10 usuarios regresan en la primera semana**
- La retención se mantiene en semanas posteriores

👉 Esto sugiere que:
- El producto tiene **valor percibido**
- Los usuarios no abandonan inmediatamente

---

## 🔄 Comportamiento por cohortes

### 🔹 Cohortes más fuertes
- **Febrero (2025-02)**:
  - Mejora en semana 3 (43.98%)
- **Marzo (2025-03)**:
  - Mejor retención en semana 2 (43.09%)

👉 Indican posibles mejoras en producto o adquisición en esos periodos

---

### 🔹 Cohortes más débiles
- **Mayo (2025-05)**:
  - Caída en semana 2 (40.07%)

👉 Puede indicar:
- Problemas recientes en experiencia
- Usuarios de menor calidad adquiridos

---

## 📊 Evolución de la retención

- No hay una caída progresiva fuerte (lo cual sería negativo)
- En algunos casos, la retención incluso aumenta en semanas posteriores

👉 Posibles explicaciones:
- Usuarios que tardan en adoptar el servicio
- Uso no inmediato pero recurrente

---

## 🧠 Insight clave de negocio

- La retención es **moderada pero estable (~42%)**
- El mayor reto no es retener, sino:
  - **Aumentar la activación inicial**
  - Convertir más usuarios en recurrentes

---

# 🚀 Oportunidades de mejora

- Mejorar onboarding del usuario (primeros días clave)
- Incentivar la segunda compra (promociones, recordatorios)
- Analizar diferencias entre cohortes (qué cambió en Feb/Mar)

---

# ✅ Conclusión del Paso 4

- Los usuarios sí regresan, pero solo ~40% lo hace  
- La retención es estable, lo cual es positivo  
- Existe oportunidad clara de crecimiento en engagement  

👉 Incrementar la retención tendría un impacto directo en:
- Revenue  
- Lifetime Value (LTV)  
- Rentabilidad del negocio  

In [16]:
# 🔹 PASO 4 - EVALUAR SI LOS USUARIOS REGRESAN (RETENCIÓN POR COHORTES)
# Explorar tabla users
# =========================
query_users = '''
SELECT *
FROM users
LIMIT 10;
'''
users = pd.read_sql(query_users, con=engine)
users.head(10)

,id_usuario,fecha_registro,país,dispositivo,tipo_plan
0,user_0,2025-01-29,Mexico,mobile,free
1,user_1,2025-01-07,Mexico,mobile,free
2,user_2,2025-03-12,Argentina,mobile,free
3,user_3,2025-03-04,Mexico,desktop,free
4,user_4,2025-02-27,Argentina,desktop,free
5,user_5,2025-02-05,Colombia,desktop,free
6,user_6,2025-01-27,Mexico,mobile,free
7,user_7,2025-05-20,Colombia,mobile,free
8,user_8,2025-01-23,Mexico,mobile,free
9,user_9,2025-04-19,Colombia,desktop,free


In [17]:
# Explorar tabla user_activity
# =========================
query_user_activity = '''
SELECT *
FROM user_activity
LIMIT 10
'''
user_activity = pd.read_sql(query_user_activity, con=engine)
user_activity.head(10)

,id_usuario,fecha_actividad,dias_despues_registro,activo
0,user_0,2025-02-05,7,0
1,user_0,2025-02-12,14,1
2,user_0,2025-02-19,21,1
3,user_0,2025-02-26,28,0
4,user_1,2025-01-14,7,0
5,user_1,2025-01-21,14,0
6,user_1,2025-01-28,21,1
7,user_1,2025-02-04,28,0
8,user_2,2025-03-19,7,0
9,user_2,2025-03-26,14,1


In [18]:
# Retención por cohortes
# ======================

query_cohort_retention_final = '''
-- Convertir fecha de registro a formato DATE
WITH users_clean AS (
  SELECT
  id_usuario,
  CAST(fecha_registro AS DATE) AS fecha_registro
  FROM users
),
-- Definir cohorte por mes de registro
cohortes AS (
    SELECT
        id_usuario,
        TO_CHAR(DATE_TRUNC('month', fecha_registro), 'YYYY-MM') AS cohorte
    FROM users_clean
),
-- Actividad de usuarios con semanas desde registro
actividad AS (
    SELECT
        ua.id_usuario,
        c.cohorte,
        ua.dias_despues_registro,
        CASE
            WHEN ua.dias_despues_registro BETWEEN 1 AND 7 THEN 'w1'
            WHEN ua.dias_despues_registro BETWEEN 8 AND 14 THEN 'w2'
            WHEN ua.dias_despues_registro BETWEEN 15 AND 21 THEN 'w3'
        END AS semana
    FROM user_activity ua
    JOIN cohortes c
        ON ua.id_usuario = c.id_usuario
    WHERE ua.activo = 1
),
-- Usuarios retenidos por cohorte y semana
retencion AS (
    SELECT
        cohorte,
        semana,
        COUNT(DISTINCT id_usuario) AS usuarios_retenidos
    FROM actividad
    WHERE semana IS NOT NULL
    GROUP BY cohorte, semana
),
-- Tamaño inicial de cada cohorte
cohorte_size AS (
    SELECT
        cohorte,
        COUNT(DISTINCT id_usuario) AS usuarios_iniciales
    FROM cohortes
    GROUP BY cohorte
)

-- Resultado final con métricas de retención
SELECT
    r.cohorte,
    cs.usuarios_iniciales,

    -- Retenidos
    SUM(CASE WHEN r.semana = 'w1' THEN r.usuarios_retenidos ELSE 0 END) AS retenido_w1,
    SUM(CASE WHEN r.semana = 'w2' THEN r.usuarios_retenidos ELSE 0 END) AS retenido_w2,
    SUM(CASE WHEN r.semana = 'w3' THEN r.usuarios_retenidos ELSE 0 END) AS retenido_w3,

    -- Porcentajes de retención
    ROUND(SUM(CASE WHEN r.semana = 'w1' THEN r.usuarios_retenidos ELSE 0 END) * 100 / cs.usuarios_iniciales, 2) AS semana_1,
    ROUND(SUM(CASE WHEN r.semana = 'w2' THEN r.usuarios_retenidos ELSE 0 END) * 100 / cs.usuarios_iniciales, 2) AS semana_2,
    ROUND(SUM(CASE WHEN r.semana = 'w3' THEN r.usuarios_retenidos ELSE 0 END) * 100 / cs.usuarios_iniciales, 2) AS semana_3

FROM retencion r
JOIN cohorte_size cs
    ON r.cohorte = cs.cohorte

GROUP BY r.cohorte, cs.usuarios_iniciales
ORDER BY r.cohorte;
'''

# Ejecutar la consulta
cohorte_final = pd.read_sql(query_cohort_retention_final, con=engine)
cohorte_final

,cohorte,usuarios_iniciales,retenido_w1,retenido_w2,retenido_w3,semana_1,semana_2,semana_3
0,2025-01,1627,697.0,668.0,656.0,42.84,41.06,40.32
1,2025-02,1444,611.0,609.0,635.0,42.31,42.17,43.98
2,2025-03,1636,677.0,705.0,690.0,41.38,43.09,42.18
3,2025-04,1606,680.0,697.0,663.0,42.34,43.40,41.28
4,2025-05,1687,695.0,676.0,706.0,41.20,40.07,41.85


---

## 🔹 Paso 5: Validar si los cambios generan impacto (test estadístico)

🎯 **Objetivo:** Evaluar si la modificación en la UI del checkout impacta la **tasa de conversión de compra**.

---

1. **Analizar el dataset** `experiment_checkout_ui.csv` para identificar la métrica principal **conversion**.
   - La métrica **conversion**
     - 1 → Usuario completó la compra  
     - 0 → Usuario no completó la compra      
2. **Plantear la hipótesis estadística**     
3. **Aplicar el test estadístico adecuado**
4. **Interpretar el resultado**


---

# 🧪 Hipótesis estadística

- **H₀ (Hipótesis nula):**  
  La tasa de conversión es igual en las variantes de las páginas control y tratamiento. Cualquier diferencia se debe al azar del muestreo.

- **H₁ (Hipótesis alternativa):**  
  La tasa de conversión es diferente en las páginas control y tratamiento. Hay un factor que influye en la decisión de compra.

---

# ⚙️ Test estadístico aplicado

- **Prueba:** z-test de proporciones  
- **Tipo de variable:** binaria (conversión: sí/no)  
- **Nivel de significancia:** α = 0.05  

---

# 📈 Resultados del experimento

- **Estadístico z:** -0.8133  
- **Valor p:** 0.4161  

### Tasas de conversión:

- **Control:** 15.69%  
- **Tratamiento:** 16.29%  

---

# 🔍 Interpretación estadística

- El **valor p (0.4161)** es mayor que α (0.05)
- **Interpretación práctica:**
  - Valor p bajo → Hay evidencia estadística contra la hipótesis nula.
  - Valor p alto → No hay evidencia suficiente para afirmar una diferencia en conversión

👉 Esto significa que:

❌ **No se rechaza la hipótesis nula**

- No hay evidencia estadísticamente significativa  
- La diferencia observada puede explicarse por el azar  

---

# 🧠 Interpretación de negocio

- Aunque la variante **tratamiento tiene una conversión ligeramente mayor (+0.6%)**,  
  esta diferencia **no es significativa**

👉 En términos prácticos:
- No se puede afirmar que el nuevo diseño mejora la conversión  
- El cambio en la UI **no genera un impacto claro en el negocio**

---

# 📌 Insight clave

- El experimento muestra una mejora leve, pero no confiable  
- Implementar el cambio implicaría riesgo sin evidencia sólida  

---

# 🚀 Recomendaciones

- No implementar el cambio basado en este experimento  
- Diseñar nuevas pruebas:
  - Cambios más significativos en la UI
  - Mejoras en el proceso de pago (detectado en el funnel)

---

# ✅ Conclusión del Paso 5

- No hay evidencia estadística de que el cambio en la UI impacte la conversión  
- La diferencia observada es pequeña y no significativa  
- Se recomienda iterar el experimento antes de tomar decisiones  

👉 Las decisiones deben ser **data-driven**, no basadas en diferencias aparentes.


In [19]:
# 🔹 PASO 5 - VALIDAR SI LOS CAMBIOS GENERAN IMPACTO (TEST ESTADÍSTICO)
checkout = pd.read_csv('https://practicum-content.s3.amazonaws.com/datasets/experiment_checkout_ui.csv')

In [20]:
# Explorar dataset
# Explorar la estructura de checkout
display("Checkout crudo: vista general de la tabla", checkout.head())
print(f"\n")
display("Checkout crudo: información general del dataset", checkout.info())
print(f"\n")
# Explorar columnas numéricas de experimento
display("Checkout crudo: estadísticas de columnas numéricas", checkout.describe())
print(f"\n")
# Explorar columnas categóricas de experimento
categoricas_checkout = ['id_usuario','variante','dispositivo','pais']
for col in categoricas_checkout:
  print(f"\n")
  display("Checkout crudo: estadísticas de columnas categóricas",checkout[col].value_counts())

'Checkout crudo: vista general de la tabla'

,id_usuario,variante,convirtio,dispositivo,pais,duracion_sesion,timestamp
0,exp_user_0,tratamiento,0,mobile,Argentina,114.41,2025-03-28
1,exp_user_1,tratamiento,0,desktop,Mexico,170.03,2025-01-15
2,exp_user_2,control,1,mobile,Colombia,140.21,2025-03-18
3,exp_user_3,tratamiento,0,mobile,Colombia,151.45,2025-06-03
4,exp_user_4,tratamiento,0,desktop,Mexico,299.96,2025-01-12




<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10000 entries, 0 to 9999
Data columns (total 7 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   id_usuario       10000 non-null  object 
 1   variante         10000 non-null  object 
 2   convirtio        10000 non-null  int64  
 3   dispositivo      10000 non-null  object 
 4   pais             10000 non-null  object 
 5   duracion_sesion  10000 non-null  float64
 6   timestamp        10000 non-null  object 
dtypes: float64(1), int64(1), object(5)
memory usage: 547.0+ KB


'Checkout crudo: información general del dataset'

None

'Checkout crudo: estadísticas de columnas numéricas'

,convirtio,duracion_sesion
count,10000.000000,10000.000000
mean,0.159900,159.862439
std,0.366532,81.074410
min,0.000000,20.010000
25%,0.000000,89.470000
50%,0.000000,159.725000
75%,0.000000,229.745000
max,1.000000,300.000000


'Checkout crudo: estadísticas de columnas categóricas'

,count
id_usuario,
exp_user_9983,1
exp_user_9982,1
exp_user_9981,1
exp_user_9980,1
exp_user_9979,1
...,...
exp_user_4,1
exp_user_3,1
exp_user_2,1


'Checkout crudo: estadísticas de columnas categóricas'

,count
variante,
tratamiento,5035
control,4965


'Checkout crudo: estadísticas de columnas categóricas'

,count
dispositivo,
desktop,5042
mobile,4958


'Checkout crudo: estadísticas de columnas categóricas'

,count
pais,
Mexico,3405
Argentina,3317
Colombia,3278


In [21]:
# Conversión de fechas
checkout['timestamp'] = pd.to_datetime(checkout['timestamp'], errors='coerce')

In [22]:
# Verificar cambio a tipo de dato fecha
display("Checkout clean: información general del dataset", checkout.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10000 entries, 0 to 9999
Data columns (total 7 columns):
 #   Column           Non-Null Count  Dtype         
---  ------           --------------  -----         
 0   id_usuario       10000 non-null  object        
 1   variante         10000 non-null  object        
 2   convirtio        10000 non-null  int64         
 3   dispositivo      10000 non-null  object        
 4   pais             10000 non-null  object        
 5   duracion_sesion  10000 non-null  float64       
 6   timestamp        10000 non-null  datetime64[ns]
dtypes: datetime64[ns](1), float64(1), int64(1), object(4)
memory usage: 547.0+ KB


'Checkout clean: información general del dataset'

None

In [23]:
# Importar libreria statsmodels
from statsmodels.stats.proportion import proportions_ztest

# Número de usuarios convertidos por página
conversiones = checkout.groupby('variante')['convirtio'].sum()

# Total de usuarios por página
totales = checkout.groupby('variante')['convirtio'].count()

display("Usuarios convertidos por página:", conversiones)
print(f"\n")
display("Total de usuarios por página:", totales)

'Usuarios convertidos por página:'

,convirtio
variante,
control,779
tratamiento,820


'Total de usuarios por página:'

,convirtio
variante,
control,4965
tratamiento,5035


In [24]:
# Aplicar prueba
exitos = [conversiones['control'], conversiones['tratamiento']]
observaciones = [totales['control'], totales['tratamiento']]
z_stat, p_value = proportions_ztest(exitos , observaciones)
tasa_control = exitos[0] / observaciones[0]
tasa_tratamiento = exitos[1] / observaciones[1]
# Visualizar resultados
print(f"Estadístico : {z_stat}")
print(f"Valor p: {p_value}")
alpha = 0.05  # umbral de significancia

if p_value < alpha:
    print("Rechazamos la hipótesis nula: hay evidencia de una diferencia.")
else:
    print("No rechazamos la hipótesis nula: no hay evidencia suficiente de una diferencia.")

print(f"\nTasa de conversión página Control: {tasa_control:.2%}")
print(f"Tasa de conversión página Tratamiento: {tasa_tratamiento:.2%}")

Estadístico : -0.8132782986429474
Valor p: 0.41605851639119995
No rechazamos la hipótesis nula: no hay evidencia suficiente de una diferencia.

Tasa de conversión página Control: 15.69%
Tasa de conversión página Tratamiento: 16.29%


---

## 🔹 Paso 6: Comunicar los resultados (Dashboard en BI)

🎯 **Objetivo**:  
Crear un dashboard que muestre de manera clara y visual los resultados del análisis de ventas, costos, marketing y conversión.

Se usarán los CSVs limpios del Paso 1:

- `orders_clean.csv`  
- `catalog_clean.csv`  
- `marketing_clean.csv`

---

## 🧩 Estructura del Dashboard

El dashboard se divide en **dos secciones principales**:

### 1️⃣ Overview Ejecutivo  
### 2️⃣ Análisis Detallado (Drill-through)

---

## 1️⃣ Overview Ejecutivo

### 📊 KPIs principales

El dashboard presenta indicadores clave en la parte superior:

- **Revenue total:** $9.49 millones

- **Profit total:** $2.84 millones

- **Gasto en marketing:** $2.87 millones

- **Ticket promedio:** $386

- **Cantidad promedio por orden:** 1.51  

👉 Permiten entender rápidamente la salud del negocio.

---

### 📈 Evolución del negocio

#### 🔹 Revenue y Profit mensual
- Gráfico de líneas muestra la evolución de ingresos y ganancias por mes (2025)
- Se observa:
  - Comportamiento estable
  - Sin caídas bruscas
  - Negocio consistente en el tiempo

---

#### 🔹 Revenue YTD y Profit YTD
- Gráfico acumulado (Year-To-Date)
- Muestra crecimiento progresivo mes a mes

👉 Insight:
- El negocio crece de forma sostenida
- No hay señales de desaceleración

---

### 📦 Desempeño por categoría

#### 🔹 Gross Profit y Revenue por categoría

- Categorías analizadas:
  - Hogar  
  - Electrónica  
  - Moda  

👉 Insights:
- Todas las categorías son rentables  
- No hay categorías con pérdidas  
- Ingresos distribuidos de forma equilibrada  

---

### 📢 Marketing

#### 🔹 Gasto por canal

- Canales:
  - Social  
  - Organic  
  - Paid Search  

👉 Insights:
- Inversión balanceada entre canales  
- No hay dependencia de un solo canal  
- Estrategia diversificada  

---

## 🧠 Insights clave del Overview

- El negocio es **rentable y estable**  
- Existe crecimiento acumulado positivo (YTD)  
- Las categorías mantienen buen desempeño  
- El marketing está distribuido de forma equilibrada  

---

## 2️⃣ Análisis Detallado (Drill-through)

### 📊 Ventas por producto

- Gráfico de barras muestra la cantidad vendida por producto  

👉 Insights:
- Productos más vendidos:
  - Vacuum-Pro-Black  
  - Jacket-Winter-M  
  - Blender-XL-Red  

---

### 📋 Tabla detallada de órdenes

Incluye:

- Producto  
- Cantidad  
- Revenue  
- Costo total  
- Gross Profit  

#### 🎨 Formato condicional:
- Verde → alta rentabilidad  
- Amarillo → rentabilidad media  
- Rojo → pérdida  

---

### 🔍 Hallazgos clave

#### 🟢 Productos más rentables
- Vacuum-Pro-Black  
- Sneakers-Urban-42  

👉 Alto revenue + bajo costo

---

#### 🟡 Productos con margen medio
- Blender-XL-Red  
- Jacket-Winter-M  

---

#### 🔴 Producto con pérdida
- Laptop-Gaming-16GB  

👉 Insight crítico:
- Alto costo unitario afecta rentabilidad  
- Requiere revisión (precio o proveedor)

---

### 🎛️ Filtros interactivos

El dashboard incluye segmentación por:

- Mes  
- País  
- Categoría de producto  
- Fuente de adquisición  

👉 Permite análisis dinámico y exploración profunda  

---

## 🚀 Valor del Drill-through

Permite:

- Pasar de visión detalle → detalle producto
- Seleccionar productos específicos mediante el clic secundario y dando en "Obtener detalles" para ver todos los pedidos relacionados
- Identificar oportunidades de optimización  

---

## 🧠 Insights globales del dashboard

El dashboard permite identificar:

- Qué productos generan más ingresos  
- Qué productos generan más ganancias  
- Dónde existen pérdidas  
- Cómo evoluciona el negocio en el tiempo  
- Cómo se distribuye el gasto en marketing  

---

## ⚠️ Consideración importante

- El **profit por producto** se analiza como **Gross Profit**  
- Esto se debe a que el marketing no está atribuido a nivel de producto  

👉 Evita distorsiones en el análisis  

---

## ✅ Conclusión del Paso 6

El dashboard desarrollado permite:

- Traducir datos en decisiones de negocio  
- Identificar oportunidades de mejora  
- Analizar rentabilidad a diferentes niveles  
- Comunicar insights de forma clara y visual  

👉 Se convierte en una herramienta clave para la toma de decisiones estratégicas.

---

## 🏁 Conclusión final del proyecto

RappiPlus es un negocio:

- ✅ Rentable  
- 📈 En crecimiento  
- ⚖️ Con buena distribución de marketing  
- ⚠️ Con oportunidades de mejora en:
  - Checkout (funnel)
  - Retención
  - Productos con bajo margen  

👉 El análisis permite pasar de datos a decisiones concretas de negocio.

---

## 🚀 Entrega Final

- 🔗 A continuación se comparte mediante links de **Google Drive** acceso al archivo (`.pbix`) del Dashboard realizado en **Power BI**  y los 3 csv limpios.

### 📎 Enlace del Dashboard

- https://drive.google.com/file/d/17Mq9M7zs0p6qx74si-SzWF9Ryp8GzZg2/view?usp=sharing

### 📎 Enlaces de los archivos csv limpios

- `catalog_clean.csv` https://drive.google.com/file/d/1Ao6zsfZ-JmC2Y2mVHkcRbFv6Yn5iy0WP/view?usp=drive_link
- `marketing_clean.csv` https://drive.google.com/file/d/1gYfrQUwDSLVYLNaB9sqjZYByuDHxMZwB/view?usp=sharing
- `orders_clean.csv` https://drive.google.com/file/d/1sx2UG_oqoLfYe2B4Fw1y0W0mujmxoPpJ/view?usp=sharing